packages/trimerge-sync/src/lib/LeaderManager.ts

```diff
@@ -0,0 +1,202 @@
+import { LeaderEvent } from '../types';
+
+type Timeout = ReturnType<typeof setTimeout>;
+type Interval = ReturnType<typeof setInterval>;
+
+export type LeaderSettings = Readonly<{
+  electionTimeoutMs: number;
+  heartbeatMs: number;
+  heartbeatTimeoutMs: number;
+}>;
+
+const DEFAULT_SETTINGS: LeaderSettings = {
+  electionTimeoutMs: 200,
+  heartbeatMs: 1_000,
+  heartbeatTimeoutMs: 2_500,
+};
+
+/**
+ * This class takes a best-effort approach to picking a single leader amongst
+ * multiple clients assuming a shared broadcast channel.
+ *
+ * Whenever a client comes online, it requests an election.
+ *
+ * 1. If there is already a leader, it will announce itself, cancelling the election.
+ * 2. Otherwise, everyone announces themselves as available.
+ *
+ * After an election timeout, the client with the alphanumerically smallest
+ * clientId announces itself as the leader to everyone.
+ *
+ * If a leader receives an announcement from another leader, a re-election will
+ * occur, and all but one of them will be demoted back to non-leader.
+ */
+export class LeaderManager {
+  private closed = false;
+  private currentLeaderId?: string = undefined;
+  private isLeader: boolean = false;
+  private potentialLeaders = new Set<string>();
+  private electionTimeout?: Timeout = undefined;
+  private leaderHeartbeat?: Interval = undefined;
+  private heartbeatTimeout?: Timeout = undefined;
+
+  constructor(
+    /**
+     * A unique identifier for each client. This can simply be a uuid
+     */
+    private readonly clientId: string,
+    /**
+     * A callback for when this client gains or loses leadership
+     */
+    private readonly onLeaderChange: (isLeader: boolean) => void,
+    /**
+     * A callback for when this class needs to broadcast messages to other clients
+     */
+    private readonly broadcastEvent: (event: LeaderEvent) => void,
+    private readonly settings: LeaderSettings = DEFAULT_SETTINGS,
+  ) {
+    this.elect();
+  }
+
+  private elect() {
+    if (this.electionTimeout) {
+      return;
+    }
+    this.potentialLeaders.clear();
+    this.potentialLeaders.add(this.clientId);
+    this.broadcastEvent({
+      type: 'leader',
+      action: 'request',
+      clientId: this.clientId,
+    });
+    this.electionTimeout = setTimeout(
+      () => this.finishElection(),
+      this.settings.electionTimeoutMs,
```

Comment:
Are these setting expected to be the same for every client that participates in the election? Do they need to be?

Reply:
yea, they should be the same… what are you thinking?

Reply:
I don't think we need to change anything here - but maybe it should be documented somewhere (if it's not already) that the settings should be the same for all clients.